### ISSUE TITLE
Proposal: Evolution

### ISSUE THREAD

**Issue by fahad19** [frintjs/frint#415]
Exploration on various aspects of the project so it can evolve, for the better.

(May get updated upon receiving further feedback)

- [Observation](#observation)
- [Proposed direction](#proposed-direction)
- [Examples of possible new direction](#examples-of-possible-new-direction)
  - [Rendering](#rendering)
    - [With `@frint/react-dom`](#with-frintreact-dom)
    - [With ReactDOM](#with-reactdom)
  - [Higher-order component](#higher-order-component)
    - [With RxJS](#with-rxjs)
  - [Dependency Injection](#dependency-injection)
    - [With React](#with-react)
  - [Props composition](#props-composition)
    - [Current state](#current-state)
    - [Subscription](#subscription)
    - [With RxJS](#with-rxjs-1)
    - [With React](#with-react-1)
  - [Root Apps](#root-apps)
  - [Blocks](#blocks)
    - [Set up `window.blocks`](#set-up-windowblocks)
    - [Render Regions](#render-regions)
    - [Register a Block](#register-a-block)
  - [Blocks and props from Region](#blocks-and-props-from-region)
  - [Accessing Blocks](#accessing-blocks)
    - [Synchronously](#synchronously)
    - [Promised](#promised)
    - [Callbacks](#callbacks)
    - [Waiting limit](#waiting-limit)
  - [Block to Block communication](#block-to-block-communication)
  - [Cascaded and scoped providers](#cascaded-and-scoped-providers)
- [Considerations](#considerations)
  - [Pros](#pros)
  - [Cons](#cons)
    - [Note about Travix](#note-about-travix)

# Observation

From the perspective of a regular JavaScript developer:

* One may not want the concept of **Apps** when doing UI development
* **Dependency Injection** is not a common use case
* **Child Apps** is not a common use case
* Loading **Region** component for any kind of React usage is not ideal
* Giving up control of React's **rendering to DOM** may discourage some
* **RxJS**-first approach is not helping beginners
* One may want to use only **certain features** of FrintJS without having to buy rest of the features

# Proposed direction

To tackle some of the problems, we can aim to be:

* Less of a framework, and more of a **suite of libraries**
* Allow selective **composition**
* Don't **lock** developers in with opinions
* Be flexible and allow easier **interop** with other libraries (like RxJS)
* Keep serving **minimum building blocks**, allowing other tools to be built around us

# Examples of possible new direction

Key differences:

* Providers data structure defined by the developer (not opinionated)
* `App` is not a requirement
* DI is not a requirement
* RxJS is not a requirement
* Component does not need to be defined in DI Container
* Taking over React's rendering to DOM is not a requirement
* No need to load Child Apps' features

### With `@frint/react-dom`

Either render with `render` for ease of use:

```js
import { render } from '@frint/react-dom';

render(
  MyComponent,
  document.getElementById('root'),

  // third argument is providers data
  {
    foo: 'foo value',
    bar: 'bar value',
  },
);
```

The last argument can also be a function, returning providers data:

```js
render(
  Component,
  domNode,

  () => ({
    foo: 'foo value',
    bar: 'bar value',
  }),
)
```

### With ReactDOM

OR render manually with React's new Context API:

```js
import {
  // created with `React.createContext()`
  FrintContext
} from '@frint/react';
import { render } from 'react-dom';

render(
  <FrintContext.Provider value={{
    foo: 'foo value',
    bar: 'bar value',
  }}>
    <MyComponent />
  </FrintContext.Provider>,

  document.getELementById('root'),
);
```

Can also happen without exposing the React Context directly:

```js
import {
  // regular Component
  FrintProvider
} from '@frint/react';
import { render } from 'react-dom';

render(
  <FrintProvider providers={{
    foo: 'foo value',
    bar: 'bar value',
  }}>
    <MyComponent />
  </FrintProvider>,

  document.getELementById('root'),
);
```

## Higher-order component

```js
import React from 'react';

import {
  // higher-order component
  // capable of emitting props
  // while having access to providers
  hoc,
} from '@frint/react';

function MyComponent(props) {
  return (
    <div>
      <p>Hello World</p>
    </div>
  );
}

const EnhancedComponent = hoc({
  // regular props generation,
  // called only once
  props({ foo, bar }, cb) => {
    // props via callback
    cb({ foo: 1 });
    cb({ foo: 2 });

    // optional unmount handler
    return () => console.log('unmounted');
  },

  // optional: access props passed from parent,
  // called as many times new props are passed
  parentProps(parentProps, providers) => {
    return parentProps;
  },

  // optional: default props (sync),
  // called only once
  defaultProps({ foo, bar }) {
    return {};
  },

  // optional: merge all of them,
  // called as many times as needed
  mergeProps({ props, parentProps, defaultProps }) {
    return {
      ...defaultProps,
      ...props,
      ...parentProps,
    };
  },

  // optional: should emit props for updating?
  // called as many times as needed
  shouldUpdate(prevProps, mergedProps, providers) {
    return true;
  },
})(MyComponent);
```

### With RxJS

```js
import React from 'react';
import { hoc } from '@frint/react-rx'; // separate package

function MyComponent(props) {
  return (
    <div>
      <p>Hello World</p>
    </div>
  );
}

const EnhancedComponent = hoc(({ providers, parentProps$ }) => {
  return Observable.of({}); // or plain object
})(MyComponent);
```

## Dependency Injection

DI is completely optional, and a standalone library:

```js
import { createContainer, resolveAndProvide } from '@frint/di';

const Container = createContainer([
  {
    name: 'foo',
    useValue: 'foo value',
  },
  {
    name: 'bar',
    useFactory({ foo }) {
      return 'bar value',
    },
    deps: ['foo'],
  }
]);

const providers = resolveAndProvide(Container);
console.log(providers.foo); // `foo value`
console.log(providers.bar); // `bar value`
```

### With React

To use the resolved providers when rendering:

```js
import { render } from '@frint/react';

render(
  Component,
  domNode,

  // function that will be called to return providers object
  resolveAndProvide(Container),
);
```

## Props composition

```js
import {
  compose, // higher-order function
  withDefaults,
  withStore,
  shouldUpdate
} from '@frint/props';

const propsComposition = compose(
  withDefaults({
    foo: 'foo value',
  }),
  withStore(
    mapState,
    mapDispatch
  ),
  shouldUpdate(props => true),
)(providers = {});
```

### Current state

Current state of props (sync):

```js
const props = propsComposition.getValue();
```

### Subscription

Regular subscription (async):

```js
const unsubscribe = propsComposition.subscribe(
  props => console.log(props)
);

unsubscribe();
```

### With RxJS

Interop:

```js
const props$ = Observable.from(propsComposition);
```

### With React

```js
import React from 'react';
import {
  withDefaults,
  withStore,
  shouldUpdate,
} from '@frint/props';

import {
  compose // higher-order component
} from '@frint/props-react';

function MyComponent(props) {
  return <p></p>;
}

export default compose(
  withDefaults(...),
  withStore(...),
  shouldUpdate(...),
)(MyComponent);
```

## Root Apps

There is nothing called Root Apps any more.

## Blocks

The new "Blocks" are the old "Child Apps".

Blocks:

* Have lifecycle hooks (`initialize` and `beforeDestroy`)
* They can be registered dynamically
* Rendered in certain places via predefined `<Region>`s
* Can have its own set of providers (with access to those cascaded from top)
* Primary app never needs to know about them early

### Set up `window.blocks`

```js
import { createContainer, resolveAndProvide } from '@frint/di';
import { BlocksRegistry } from '@frint/blocks';

const Container = createContainer([
  {
    name: 'foo',
    useValue: 'foo value',
  },
  {
    name: 'bar',
    useValue: 'bar value',
  },
  {
    name: 'blocks',
    useFactory({ container }) {
      return new BlocksRegistry(container, {
        cascade: ['foo', 'bar'],
      });
    },
    deps: ['container'],
  },
]);
const providers = resolveAndProvide(Container);

window.blocks = providers.blocks;
```

### Render Regions

```js
import React from 'react';
import { render } from '@frint/react-dom';
import { Region } from '@frint/blocks-react';

function Root() {
  return (
    <div>
      <Region name="sidebar" />
    </div>
  );
}

render(
  <Root />,
  document.getElementById('root'),
  providers,
);
```

### Register a Block

```js
function BlockComponent() {
  return <p>Hello World from Block.</p>;
}

// register
window.blocks.register({
  name: 'Block1',
  provide(cascadedProviders = {}) {
    return {
      baz: 'baz value',
    };
  },
  initialize() {
    // can access `this.providers`
  },
  beforeDestroy() { },
  component: BlockComponent,
  regions: ['sidebar'],
  multi: false, // multi-instance: optional, defaults to `false`
});
```

## Blocks and props from Region

`RegionService` is dropped. Functionality lives in React-land.

Regular usage:

```js
import React from 'react';
import { withRegionProps } from '@frint/blocks-react';

function BlockComponent(props) {
  const { regionData } = props;

  return <p></p>;
}

export default withRegionProps((regionProps, providers) => {
  return (
    <BlockComponent regionData={regionProps.data} />
  );
});
```

## Accessing Blocks

Registering new Blocks:

```js
// first Block
window.blocks.register({
  name: 'FirstBlock',
  provide({ blocks }) {
    return {
      blocks,
      foo: 'foo value',
    };
  },
  component: Block1Component,
  regions: ['sidebar'],
});

// second Block
window.blocks.register({
  name: 'SecondBlock',
  provide({ blocks }) { return {}; },
  component: Block2Component,
  regions: ['footer'],
});
```

Access Block:

### Synchronously

```js
// Block instance or null
const block = window.blocks.getSync('FirstBlock');
```

### Promised

```js
// no cancellation - dangerous for memory leaks
window.blocks.get('FirstBlock').then(block => {
  console.log(block.name); // `FirstBlock`
  console.log(block.providers); // `{ blocks, foo }`
});
```

### Callbacks

```js
// with cancellation for listening
const unlisten = window.blocks.get('FirstBlock', (error, block) => {
  // emits once
  console.log(block.name);
});
unlisten();
```

### Waiting limit

```js
// waiting limit (Promise - no cancellation)
window.blocks.waitFor('FirstBlock', 2000)
  .then(block => console.log(block));
  .catch(() => console.log('Not available within 2 seconds'));

// waiting limit (callback - with cancellation)
const unlisten = window.blocks.waitFor('FirstBlock', 2000, (error, block) => {
  if (error) {
    console.log('Not available within 2 seconds');
    return;
  }

  console.log(block);
});
unlisten();
```

## Block to Block communication

```js
// FirstBlock
import React from 'react';
import { hoc } from '@frint/react';

function FirstBlockComponent(props) {
  const { secondBlockState } = props;

  return <p></p>;
}

export default hoc({
  props({ blocks }, cb) => {
    let unsubscribeBlockStore = null;

    const unsubscribeBlock = blocks.waitFor(
      'SecondBlock',
      2000,
      (error, block) => {
        unsubscribeBlockStore = block.providers.store.subscribe(
          state => cb({
            secondBlockState: state,
          })
        );
      });

    return () => {
      unsubscribeBlock();
      unsubscribeBlockStore && unsubscribeBlockStore();
    };
  }
})(FirstBlockComponent);
```

## Cascaded and scoped providers

We currently have ["cascaded scoped providers"](https://frint.js.org/docs/packages/frint#scoped), which allow you to set a Provider from Root App, and have it cascaded to all the Child Apps too, but the classes will be instantiated again with the App in scope (Child App).

We can achieve something similar by doing this:

```js
class Foo {
  constructor(args) {
    const { block } = args;
  }

  // return's a new instance of self
  // scoped by given Block
  forBlock(blockInstance) {
    return new Foo({
      block: blockInstance,
    });
  }
}
```

Render initially with `foo` in providers:

```js
import { render } from '@frint/react-dom';
import { BlocksRegistry } from '@frint/blocks';

const providers = {
  foo: new Foo (),
};

window.blocks = new BlocksRegistry(providers, {
  cascade: ['foo'],
});

render(
  Component,
  domNode,
  providers,
);
```

Register a new Block:

```js
window.blocks.register({
  name: 'MyBlock',
  // ...
  provide({ blocks, foo }) {
    // `foo` is returned value of `Foo.forBlock(myBlockInstance)`
    return {
      foo,
    };
  },
});
```

# Considerations

Before making any decision, we can take a look at the pros and cons of all these proposed changes from various perspectives.

## Pros

For all:

* This makes FrintJS more **approachable**
* The **learning curve** goes down significantly for regular React developers
* You don't have to learn about **multiple packages** at once
* You aren't forced to use **RxJS** (choose it at will)
* You don't have to load too many features in your application, unless you really need them (you have a **choice**)

Technically:

* Better API support with React's new (since v16.3):
  * **Async** rendering
  * **Context** API
  * Safer **server-side** rendering

## Cons

For new users:

* N/A

For existing users and maintainers:

* This involves a LOT of **breaking changes**
* This involves a LOT of work that needs to be done in almost all the packages that we currently have across **multiple monorepos** on GitHub
* Changes the very **core principles** of how FrintJS itself works (for e.g., no more Apps)
* **Backwards compatibility** would be quite hard to achieve
* **Migrating existing applications** would involve a significant amount of work
* After establishing all the **common practices** in last couple of years, we may be opening up a whole new wormhole

### Note about Travix

* The project is indeed open source and has been developed very publicly and openly, it was built primarily for the **needs of Travix**
* We don't have a good track record of **upgrading FrintJS** in our production websites in time (irrespective of the amount of migration work, even though the maintainers always provide detailed guides with each major semver release)
* Personally, I feel it will take **more than a year** to get all these things to happen (if they happen at all), including further development of FrintJS and rolling out the changes in production

According to the points above, I can understand if we decide **not to go** for this proposal in this project.

---

Now that most of the info is laid out in the open, we need a decision. The process of decision making is quite conflicting, since **two sides** can be involved:

1. Point of view from a **generic purpose** open source project
1. Point of view from a **company** for which this project was initialy built

I would highly appreciate all of your feedback on this.

Cheers!

**Comment by viacheslaff** [comment]
Impressive effort and high-level thinking @fahad19! Very interesting ideas.

While it's a quite consistent proposal, it's pretty big and it's hard to start discussing it. I t has many points and has both high-level and detailed ideas. I'd suggest either split it into several or focus on high-level discussion and then go into details.

If you choose splitting it can be done by discussing separately loosening each of those requirements and changes to the API (while keeping the high-level ideas in mind):

- Providers data structure defined by the developer (not opinionated)
- App is not a requirement
- DI is not a requirement
- RxJS is not a requirement
- Component does not need to be defined in DI Container
- Taking over React's rendering to DOM is not a requirement
- No need to load Child Apps' features

If you choose to focus on high-level discussion (which I'd start with), these points can be discussed:

- Less of a framework, and more of a suite of libraries
- Allow selective composition
- Don't lock developers in with opinions
- Be flexible and allow easier interop with other libraries (like RxJS)
- Keep serving minimum building blocks, allowing other tools to be built around us

Once again, great work, but for productive discussion to happen the scope should be limited, and the discussion should be guided and structured.

**Comment by andrewparadi** [comment]
@fahad19, great job reimagining a less opinionated version of FrintJS moving forward.

I certainly agree strongly with the worries around the following:
1. Huge breaking changes and forced rewriting of applications that use FrintJS make it very difficult to agree to continue development of FrintJS applications given how much upgrade work will be required in 1-2 years time.
2. Sheer scope of the project such that this next version of FrintJS could take over 1 year to transition to

Given the project I'm using FrintJS for is still in it's early days it would be frustrating to put in this effort to use the framework only to have all of it in substantial ways be broken in the next update.

Specifically to the worries expressed about rigidness and approachability, at least for our application those have been a **positive** for our development process. In building a modular dashboard, the current version of FrintJS has been instrumental in allowing us to have on demand loading of modules and tabs to an admin dashboard. I don't see the vision that initiated the project as obsolete by any means.

I think a framework as outlined above would be valuable but the migration/upgrade process should hopefully be top of mind when release approaches or should be released as a totally different framework name.

**Comment by fahad19** [comment]
Thanks for such a thoughtful response, @andrewparadi!

> I think a framework as outlined above would be valuable but the migration/upgrade process should hopefully be top of mind when release approaches or should be released as a totally different framework name.

I think that's what's gonna happen in the end. FrintJS will continue to maintain its API stability for as long as possible, while upgrading to latest ReactJS + RxJS periodically.

Other new ideas can possibly be expressed in fresh new libraries :)

**Comment by fahad19** [comment]
FrintJS will not go for any major breaking changes in near future. So closing it.
